frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeVersionDetector.java

```diff
@@ -0,0 +1,160 @@
+package com.github.eirslett.maven.plugins.frontend.lib;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Objects.isNull;
+import static java.util.Optional.empty;
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class NodeVersionDetector {
+
+    private static final String TOOL_VERSIONS_FILENAME = ".tool-versions";
+
+    public static String getNodeVersion(File workingDir, String providedNodeVersion, String genericNodeVersionFile) throws Exception {
+        Logger logger = getLogger(NodeVersionDetector.class);
+
+        if (!isNull(providedNodeVersion) && !providedNodeVersion.trim().isEmpty()) {
+            logger.debug("Looks like a node version was set so using that: " + providedNodeVersion);
+            return providedNodeVersion;
+        }
+
+        if (!isNull(genericNodeVersionFile) && !genericNodeVersionFile.trim().isEmpty()) {
+            File genericNodeVersionFileFile = new File(genericNodeVersionFile);
+            if (!genericNodeVersionFileFile.exists()) {
+                throw new Exception("The Node version file doesn't seem to exist: " + genericNodeVersionFileFile);
+            }
+
+            if (genericNodeVersionFile.endsWith(TOOL_VERSIONS_FILENAME)) {
+                return readToolVersionsFile(genericNodeVersionFileFile, genericNodeVersionFileFile.toPath(), logger);
+            } else {
+                return readNvmrcFile(genericNodeVersionFileFile, genericNodeVersionFileFile.toPath(), logger);
+            }
+        }
+
+        try {
+            return recursivelyFindVersion(workingDir);
+        } catch (Throwable throwable) {
+            logger.debug("Going to use the configuration node version, failed to find a file with the version because",
+                    throwable);
+            return providedNodeVersion;
+        }
+    }
+
+    public static String recursivelyFindVersion(File directory) throws Exception {
+        Logger logger = getLogger(NodeVersionDetector.class);
+
+        if (!directory.canRead()) {
+            throw new Exception("Tried to find a Node version file but giving up because it's not possible to read " +
+                    directory.getPath());
+        }
+
+        String directoryPath = directory.getPath();
+
+        Path nodeVersionFilePath = Paths.get(directoryPath, ".node-version");
```

Comment:
nit: Seems we could make these dry with a loop over an array of filenames and callbacks

Reply:
nit: maybe it would be worth to issue some warnings in case the values are mismatched between the 3 files (if/when more than one exists)?

Reply:
See above